### DISCUSSION TITLE
increased file size when using apply_redactions()

### DISCUSSION THREAD

**Discussion by lbr991** [pymupdf/PyMuPDF#2644] [Looking for help] [answered]
I am getting a huge increase in filesize and have pinpointed to just one line (!):
`            pdf_page.apply_redactions()
`
If I save the file without calling apply_redactions() the size is 55KB, if I do the size is 800KB. 

Why is this happening? It only draws 4 small black rectangles.

**Comment by JorjMcKie** [comment]
What were the save options you used?

**Reply by lbr991** [reply]
Just `doc.save(output)`

Regarding the annotation:
```
    rect = fitz.Rect(
        bbox['Left'] * page.rect.width, 
        bbox['Top'] * page.rect.height, 
        (bbox['Left'] + bbox['Width']) * page.rect.width, 
        (bbox['Top'] + bbox['Height']) * page.rect.height
    )
    annot = page.add_redact_annot(rect)
    annot.set_colors(stroke=(0, 0, 0), fill=(0, 0, 0))
    annot.set_opacity(1)
    annot.update()
```

**Reply by JorjMcKie** [reply, marked as answer]
> Just `doc.save(output)`

Well, this will not physically delete stuff that you only **think** you have deleted. All old objects are still there as "ghosts" and could be accessed by hackers.
Use at least "garbage=3, deflate=True", or equivalently `doc.ez_save(...)`.

**Reply by lbr991** [reply]
Thanks, it works!
Does `doc.save(output, garbage=3, deflate=True)` have the same effect as `doc.ez_save(...)`? 
Also why do you mention "at least"? Do you recommend to do anything more for redactions?

**Reply by JorjMcKie** [reply]
At least means no less than garbage=3. There are (rare) occasions, where garbage=4 compresses even better than 3: it scans binary streams too to discover equality. Such a situation should however not be caused by applying redactions.
`doc.ez_save()` (pronounce as "easy save") is just an abbreviation for the most commonly used save with reasonable compression - meaning a good balance between compression result and processing required to do it.

**Reply by lbr991** [reply]
In Adobe when the user redacts a pdf, it shows:
![image](https://github.com/pymupdf/PyMuPDF/assets/136107003/3bb6bf71-f07c-4ec5-992a-7a55aa64c986)

![image](https://github.com/pymupdf/PyMuPDF/assets/136107003/b835d7be-dcfb-4800-adab-edefaa37adcd)
https://www.adobe.com/acrobat/resources/how-to-redact-a-pdf.html

What is "hidden data and metadata" in this case? 

Is hidden data the "ghosts" handled by `garbage=3` and metadata is `title, author, subject` etc? 
Are there any other measures I can take for redacting data?

**Reply by JorjMcKie** [reply]
> Is hidden data the "ghosts" handled by garbage=3 and metadata is title, author, subject etc?

Yes, it means all the stuff removed by garbage collection in PyMuPDF.

> Are there any other measures I can take for redacting data?

Normally, there are two different organizational units that (1) enter the redaction amd (2) apply it.
Obviously, the second one could decide to reject the redaction. Or change how intersecting images should be handled. Rejecting it altogether will restore earlier data for example.
I can't image what sense it could make to **_not remove the ghosts_**.

**Reply by JorjMcKie** [reply]
Otherwise, "sanitizing" in the PDF sense means things that can be done using `Document.scrub()`.

**Reply by lbr991** [reply]
If the redaction gets rejected, does it emit a warning?
Also does it matter if the visual redaction (black rectangle) comes before the text layer redaction?

**Reply by JorjMcKie** [reply]
Rejecting a redaction, must be an organizational process or a capability offered to the rejector, such that he can simply delete the redact annot. If he does so, the PDF will be unchanged, and no black rectanle drawn.

**Reply by lbr991** [reply]
What are the reasons that `garbage=3` is preferred to `garbage=4`?

**Reply by JorjMcKie** [reply]
garbage=4 will take considerably longer, because it will read all streams and compare them to find any duplicates. Choose it only if you suspect there exist such things.

**Reply by lbr991** [reply]
Can you list an example case where a duplicate stream can exist? The way I understand this (please correct me if I'm wrong) is that if I'm prioritizing "safety" (confidence that the data has been redacted) over speed I should opt for garbage=4?

**Reply by lbr991** [reply]
@JorjMcKie Can you please reply to this?

**Comment by lbr991** [comment]
> Rejecting a redaction, must be an organizational process or a capability offered to the rejector, such that he can simply delete the redact annot. If he does so, the PDF will be unchanged, and no black rectanle drawn.

Not sure what you mean, I thought "organizational process" was something defined in the PDF format?

**Reply by JorjMcKie** [reply]
Adding a redact annot means suggesting a change.
Applying it means accepting it.
Deleting a redact annot means rejecting the change.
That's all.

**Reply by lbr991** [reply]
Just to confirm, if I add a redact annot and apply it, then use garbage=3, a malicious user would not be able to delete the redact annotation from output.pdf, correct?

**Reply by JorjMcKie** [reply]
Not quite:
After applying the readctions, their changes will be permanent part of the PDF and cannot be reverted (if not saving incrementally).
Reasonable garbage collection will recover file space by deleting objects that have become ghosts during the process. This will also prevent to access those previous data - which is important, because one of redaction motivations is data protection.
The redaction annotations themselves however **_will_** be deleted.

**Comment by JorjMcKie** [comment]
> Just to confirm, if I add a redact annot and apply it, then use garbage=3, a malicious user would not be able to delete the redact annotation from output.pdf, correct?

When you execute `page.apply_redactions()` then the redacted data and all redaction annotations are removed.
Logically at first. Physically also if saving with a sufficient garbage option.
Physically also means that nothing in the world can recover things removed by the redaction - as well as the redaction annot itself.

**Reply by lbr991** [reply]
Thanks for that clarification but meant this earlier question:

> Can you list an example case where a duplicate stream can exist? The way I understand this (please correct me if I'm wrong) is that if I'm prioritizing "safety" (confidence that the data has been redacted) over speed I should opt for garbage=4?

**Reply by JorjMcKie** [reply]
E.g. some careless PDF creator may have inserted the same image twice under different names and different xrefs. This does not happen under the control of the redaction algorithm.
But of course: who knows where that PDF has come from, it may have contained grease before you started dealing with it.

**Comment by luchux** [comment]
@JorjMcKie Would write(garbage=4, deflate=True) be expected to also clear redactions and release memory?

**Reply by luchux** [reply]
with those options  and ez_save, or save and those options, I am still seeing a significantly increase in size (from 4mb to 21mb) when adding 10 redaction annotation and applying them.

**Reply by JorjMcKie** [reply]
As part of the redaction process, images will be changed that intersect a redact rectangle - if the default option images=fitz.PDF_REDACT_IMAGE_PIXELS is used. These changes lead to converting affected images to PNG format ... but **_only_** for the page where redactions are being processed!
Some implications are:
* Previous other image formats are converted to PNG format. Depending on the previous format, a larger image size will result.
* If the same image is used on multiple pages, then these other pages will still use their original image. So file size will increase b/o of an additional image.

**_Example_**
A PDF with 4 pages all showing the same JPG image shows this initially:
```python
for page in doc:
    print(page.get_images())

[(5, 0, 439, 501, 8, 'DeviceRGB', '', 'fzImg0', 'DCTDecode')]  # all the same JPG image
[(5, 0, 439, 501, 8, 'DeviceRGB', '', 'fzImg0', 'DCTDecode')]
[(5, 0, 439, 501, 8, 'DeviceRGB', '', 'fzImg0', 'DCTDecode')]
[(5, 0, 439, 501, 8, 'DeviceRGB', '', 'fzImg0', 'DCTDecode')]
```
After applying a redaction on page 0 (intersecting image at xref 5):
```python
for page in doc:
    print(page.get_images())

[(12, 0, 439, 501, 8, 'DeviceRGB', '', 'Im1', 'FlateDecode')]  # PNG image
[(5, 0, 439, 501, 8, 'DeviceRGB', '', 'fzImg0', 'DCTDecode')]  # same JPG image for the other pages
[(5, 0, 439, 501, 8, 'DeviceRGB', '', 'fzImg0', 'DCTDecode')]
[(5, 0, 439, 501, 8, 'DeviceRGB', '', 'fzImg0', 'DCTDecode')]
```
So in spite of using best compression, the image at xref 12 is (1) larger (PNG) than the JPG original, and (2) comes on top in terms of size.
